File: frontend/src/types/canonical.ts
/**
 * Canonical data models and workflow contracts for SAMANVAYSETU.
 *
 * In accordance with the canonical-data boundary architecture:
 * Adapters transform department-specific raw envelopes into these canonical shapes.
 * Department-specific schemas never leak to workflow decisions or the UI.
 */

export interface CitizenProfile {
  citizen_id: string;
  full_name: string;
  name?: string; // alias for display
  date_of_birth: string; // ISO 8601 YYYY-MM-DD
  mobile?: string;
}

export interface EducationRecord {
  citizen_id: string;
  student_id: string;
  student_name: string;
  date_of_birth: string; // ISO 8601 YYYY-MM-DD
  course_name: string;
  course?: string; // alias for display
  institution_name: string;
  institution?: string; // alias for display
  enrollment_status: string; // e.g., "ACTIVE"
}

export interface IncomeRecord {
  citizen_id: string;
  applicant_name: string;
  annual_income: number;
  annual_family_income?: number; // alias for display
  certificate_number?: string;
  income_certificate_no?: string;
  financial_year: string;
}

export interface FieldMappingRule {
  sourceField: string;
  sourceType: string;
  canonicalModel: "CitizenProfile" | "EducationRecord" | "IncomeRecord";
  canonicalField: string;
  transformDescription?: string;
  required?: boolean;
}

export interface MappingIssue {
  type: "missing_field" | "unknown_field" | "unmapped_field";
  department: string;
  fieldName: string;
  message: string;
  severity: "error" | "warning";
}

export interface FieldTransformationTrace {
  department: string;
  sourceField: string;
  mappingVersion: string;
  canonicalModel: string;
  canonicalField: string;
  sourceValue: unknown;
  transformedValue: unknown;
}

export interface SchemaNormalizationResult<T> {
  canonical: T;
  issues: MappingIssue[];
  traces: FieldTransformationTrace[];
  mappingVersion: string;
  engineUsed: "deterministic" | "ai-assisted";
}

export interface EligibilityCriterion {
  id: string;
  name: string;
  requirement: string;
  actual: string;
  passed: boolean;
  explanation: string;
}

export interface EligibilityEvaluation {
  is_eligible: boolean;
  criteria: EligibilityCriterion[];
  summary: string;
  evaluated_at: string;
}

export interface ScholarshipEligibilityPolicy {
  maxAnnualIncome: number;
  currency: string;
  eligibleCourseKeywords: string[];
  eligibleEnrollmentStatuses: string[];
}

export const DEFAULT_SCHOLARSHIP_POLICY: ScholarshipEligibilityPolicy = {
  maxAnnualIncome: 300000,
  currency: "INR (₹)",
  eligibleCourseKeywords: ["Engineering", "Science", "Technology", "Degree", "B.Tech", "B.Sc", "M.Sc", "Computer"],
  eligibleEnrollmentStatuses: ["ACTIVE", "ENROLLED"],
};

export interface ConsentGrant {
  consent_id: string;
  citizen_id: string;
  purpose: string;
  departments: string[];
  data_categories: string[];
  granted_at: string;
  explicit_approval: boolean;
}

export type StepState = "pending" | "running" | "success" | "failed";

export interface WorkflowStep {
  id: string;
  stepNumber: number;
  label: string;
  department?: string;
  status: StepState;
  error?: string;
  startedAt?: string;
  completedAt?: string;
  details?: string;
}

export interface AuditEvent {
  id: string;
  timestamp: string;
  event_type:
    | "consent_granted"
    | "department_requested"
    | "department_response_received"
    | "normalization_completed"
    | "eligibility_evaluated"
    | "application_submitted"
    | "step_failed"
    | "step_retried";
  department?: string;
  details: Record<string, unknown>;
}

export interface CanonicalScholarshipApplication {
  application_id: string;
  application_reference: string;
  citizen_id: string;
  student_id: string;
  applicant_name: string;
  date_of_birth: string;
  course_name: string;
  institution_name: string;
  annual_income: number;
  financial_year: string;
  eligibility_status: "ELIGIBLE" | "INELIGIBLE";
  workflow_status: "DRAFT" | "PROCESSING" | "SUBMITTED" | "FAILED" | "REJECTED";
  eligibility_details: EligibilityEvaluation;
  records_verified: {
    identity: boolean;
    education: boolean;
    income: boolean;
  };
  departments_contacted: string[];
  submission_receipt?: {
    application_id: string;
    application_reference: string;
    status: string;
    received_at: string;
    source: string;
  };
  audit_events: AuditEvent[];
  created_at: string;
  completed_at?: string;
  mode: "real" | "demo";
}

File: backend/app/schemas/canonical.py
"""Canonical data models for SAMANVAYSETU interoperability boundary.

These Pydantic models define the versioned platform representation.
Adapters transform raw departmental source shapes into these canonical structures.
"""

from datetime import datetime
from typing import Any, Literal
from pydantic import BaseModel, Field


class CitizenProfile(BaseModel):
    citizen_id: str = Field(description="Unique citizen reference identifier")
    full_name: str = Field(description="Canonical normalized person name")
    date_of_birth: str = Field(description="ISO 8601 YYYY-MM-DD birth date")
    mobile: str | None = Field(default=None, description="Contact mobile number")


class EducationRecord(BaseModel):
    citizen_id: str = Field(description="Citizen identifier cross-reference")
    student_id: str = Field(description="Institution student registration id")
    student_name: str = Field(description="Normalized student name")
    date_of_birth: str = Field(description="ISO 8601 YYYY-MM-DD birth date")
    course_name: str = Field(description="Canonical degree/course name")
    institution_name: str = Field(description="Accredited institution name")
    enrollment_status: str = Field(description="Normalized enrollment status (e.g. ACTIVE)")


class IncomeRecord(BaseModel):
    citizen_id: str = Field(description="Citizen identifier cross-reference")
    applicant_name: str = Field(description="Verified taxpayer/applicant name")
    annual_income: int = Field(ge=0, description="Gross annual income in INR (₹)")
    financial_year: str = Field(description="Assessment financial year (e.g. FY 2025/26)")


class EligibilityCriterion(BaseModel):
    id: str
    name: str
    requirement: str
    actual: str
    passed: bool
    explanation: str


class EligibilityEvaluation(BaseModel):
    is_eligible: bool
    criteria: list[EligibilityCriterion]
    summary: str
    evaluated_at: datetime


class ScholarshipApplication(BaseModel):
    application_id: str = Field(description="Scholarship application reference")
    application_reference: str = Field(description="Gateway interoperability correlation reference")
    citizen_id: str
    student_id: str
    applicant_name: str
    date_of_birth: str
    course_name: str
    institution_name: str
    annual_income: int = Field(ge=0)
    financial_year: str
    eligibility_status: Literal["ELIGIBLE", "INELIGIBLE"]
    workflow_status: Literal["DRAFT", "PROCESSING", "SUBMITTED", "FAILED", "REJECTED"]
    eligibility_details: EligibilityEvaluation
    records_verified: dict[str, bool]
    departments_contacted: list[str]
    submission_receipt: dict[str, Any] | None = None
    audit_events: list[dict[str, Any]] = Field(default_factory=list)
    created_at: datetime
    completed_at: datetime | None = None
    mode: Literal["real", "demo"] = "real"
